src/index/Vocabulary.h

```diff
@@ -85,6 +85,123 @@ class PrefixComparator {
   const Vocabulary<S>* _vocab;
 };
 
+/**
+ * \brief Comparator for std::strings that optionally supports
+ * case-insensitivity.
+ *
+ * If the constructor is called with ignoreCase=false it is an ordinary string
+ * compare If ignoreCase=True the operator behaves as follows:
+ *
+ * - compare the strings according to their lowercase version
+ * - in case the lowercase versions are equal, return the order of the original
+ * strings
+ * - (including case)
+ *
+ * This gives us a strict ordering on strings.
+ */
+class StringSortComparator {
+ public:
+  static std::string rdfLiteralToValueForLT(const std::string& input) {
+    auto rhs_string = ad_utility::getUppercaseUtf8(input);
+    auto split = StringSortComparator::extractComparable(rhs_string);
+    if (split.isLiteral && !split.langtag.empty()) {
+      // get rid of possible langtags to move to the beginning of the
+      // range
+      rhs_string = '\"' + std::string(split.val) + '\"';
+    }
+    return rhs_string;
+  }
+
+  static std::string rdfLiteralToValueForGT(const std::string& input) {
+    auto rhs_string = ad_utility::getLowercaseUtf8(input);
+    auto split2 = StringSortComparator::extractComparable(rhs_string);
+    if (split2.isLiteral) {
+      rhs_string = '\"' + std::string(split2.val) + '\"' + "@" + char(127);
+    }
+    return rhs_string;
+  }
+
+  StringSortComparator(bool ignoreCase = false) : _ignoreCase(ignoreCase) {}
+
+  bool getIgnoreCase() const { return _ignoreCase; }
+
+  bool operator()(std::string_view a, std::string_view b) const {
+    if (!_ignoreCase) {
+      return a < b;
+    } else {
+      // TODO<Johannes> Ideally we want to have this also when doing
+      // case-insensitive compare, but it currently breaks the prefix
+      // compression (there we really need ordering by correct bytes)
```

Comment:
@niklas88 I could implement this also for case-insensitive sort (sorting first by "inner" literal value and only then by language tag, should I?

Reply:
The point is the following: 
`"pref!" < "pref"`, because `! < "` in ASCII. (assume all quotation marks to be escaped).
This can be circumvented by first getting the actual string value without the `""` and then sorting by this. I am doing this in the case-insensitive case, but it is also possible for case-sensitive sorting.

In this case we would also have to build two vocabularies, one for compression and one not, even if using case-sensitive search.

However, this only affects words ending with `!` since this is the only printable ASCII character that has a smaller value than `"`

Does this make more sense to you and help you to make an informed decision?

Reply:
Yes that makes sense. I'd propose we keep the case sensitive variant as is. Having it be just byte wise comparison makes it at least easier to reason about. Also adding the described behavior sounds like complication for little gain